### PR TITLE
GGML README

### PR DESCRIPTION
**Pull request by danforbes**
Rendered document can be viewed here https://github.com/danforbes/llama-rs/blob/dfo/doc/ggml/crates/ggml/README.md

### CODE REVIEW

**Comment by danforbes** [comment]
I need help with (at least) vocabulary-scoring before this is ready to come out of draft.

**Comment by philpax** [comment]
This is really well done! Awesome work, and sorry about not looking at it earlier!

Honestly, it's so thorough I'm wondering if I should encourage you to submit it to upstream GGML instead. They could certainly benefit from the documentation!

File: crates/ggml/README.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+word _fragments_. Just like humans can compose millions of words from just a
+dozen or two letters, large language models use _tokens_ to express a large
+number of words from a relatively smaller number of components. Consider a
+vocabulary with the following tokens: `whi`, `ch` `le`, `who`, and `a`; this
+vocabulary can be used to create the English words "which", "while", "who", "a",
+and "leach". How would the behavior change if the model contained the following
+tokens: `wh`, `ich`, `ile`, `o`, and `leach`? Choices such as these allow
+model-creators to tune the behavior and performance of their models.
+
+As described above, the model's [hyperparameters](#hyperparamaters) typically
```
Comment (philpax) [inline crates/ggml/README.md:72]: #hyperparameters

File: crates/ggml/README.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+contains a value that specifies the number of tokens in the vocabulary. The
+vocabulary is encoded as a list of tokens, each of which includes an unsigned
+32-bit integer that specifies the length of the token. Depending on the GGML
+version, the token may also include a 32-bit floating point score.
+
+[comment]: <> (I need help describing token-scoring)
```
Comment (philpax) [inline crates/ggml/README.md:78]: I'd suggest linking to this: https://huggingface.co/learn/nlp-course/chapter6/5

but essentially, each token is given a score / probability, and the tokenization algorithm tries to select for tokens with the maximal probability.

File: crates/ggml/README.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
+# GGML - Large Language Models for Everyone
+
+[GGML](https://github.com/ggerganov/ggml) is a format for distributing large
+language models (LLMs); the name is a portmanteau of the initials of its
```
Comment (philpax) [inline crates/ggml/README.md:4]: Suggest mentioning that GGML is a library for ML stuff first, and then mention the format. The format can also be used for any ML thing that targets GGML-the-library (including whisper.cpp), not strictly LLMs, but that's where it's taken off.

**Review by philpax — CHANGES_REQUESTED** [review]
Alright, let's merge it here! Just a few small comments, but excellent work writing this up. It'll be a major help for both current and future contributors 🚀

**Comment by danforbes** [comment]
Can you write one sentence that describes the purpose of token-scoring? Is this correct?

> Depending on the GGML version, the token may also include a 32-bit floating point score, which impacts how likely that token is to appear in text generated by the model.

**Comment by philpax** [comment]
> Can you write one sentence that describes the purpose of token-scoring? Is this correct?
> 
> > Depending on the GGML version, the token may also include a 32-bit floating point score, which impacts how likely that token is to appear in text generated by the model.

I'd say more "how likely that token is to appear in the dataset that the model was trained on", but otherwise yeah that's fine

**Review by philpax — APPROVED** [review]

**Comment by danforbes** [comment]
I'm ready for this to be merged if you are 👍🏻

**Comment by danforbes** [comment]
> Honestly, it's so thorough I'm wondering if I should encourage you to submit it to upstream GGML instead. They could certainly benefit from the documentation!

Oh, that was always part of the plan 😉 https://github.com/ggerganov/ggml/pull/177

**Pull request merged**